### PR TITLE
fix(nc): fix the determinism of the generated sources of the nodeset compiler

### PR DESCRIPTION
**Pull request by tobiajung**
Currently, the nodeset compiler creates source file including the nodes of the information models
in a different order every time a rebuild is done. This is due to the use of sets which change the
order of there elements every time an element gets added. With this change the nodeset compiler
uses lists where order is relevant.

### CODE REVIEW

**Comment by ccvca** [comment]
Related to #5000

**Comment by jpfr** [comment]
Cool! Does this have a significant impact on timings?
I ask because the lookup to test membership becomes O(n) instead of O(1).

**Comment by tobiajung** [comment]
> Does this have a significant impact on timings?

Unfortunately, generating the nodeset with this change can take up to 4x longer on my machine.
Generating the full nodeset on open62541/master takes 4.3 seconds while with my requested change it takes 16.4 seconds.
The increase in time is mainly due to the change to list() in the references member in the Node class.

**Comment by tobiajung** [comment]
OrderedDicts are key value pairs. So I don't think that they can replace the sets for the references of nodes.
Sets and lists only contain one or more elements which can be accessed through an index and not through a key.

**Comment by jpfr** [comment]
A 4x performance decrease is not possible.
So we need a smart solution.

OrderedDict can be used to create a stable set.
For example by putting the element as the key that point to a `None`.

This can also be wrapped in a small class for reuse.
See here for a more complete example:
https://github.com/rspeer/ordered-set/blob/457a1ec89660f6491ea36b05b44e55dc42f6a873/ordered_set/__init__.py#L55

**Comment by tobiajung** [comment]
Thank you very much for your detailed information! I will take a look at it next week

File: tools/nodeset_compiler/nodes.py [hunk 1]
```diff
@@ -69,7 +69,7 @@ def __init__(self):
         self.symbolicName = None
         self.writeMask = None
         self.userWriteMask = None
-        self.references = set()
+        self.references = dict()
```
Comment (jpfr) [inline tools/nodeset_compiler/nodes.py:72]: Add a short comment to document the “special” use of dict and the motivation for it.
So future generations can understand what you did.

File: tools/nodeset_compiler/nodeset.py [hunk 2]
```diff
@@ -227,16 +227,16 @@ def filterRef(r, rt):
             if r.target == node.id:
                 if r.source not in self.nodes:
                     continue
-                self.nodes[r.source].references = set(filter(
-                    lambda rt: filterRef(r, rt),
-                    self.nodes[r.source].references
+                self.nodes[r.source].references = dict(filter(
+                    lambda rt: filterRef(r, rt[0]),
```
Comment (jpfr) [inline tools/nodeset_compiler/nodeset.py:231]: Why rt[0]?

File: tools/nodeset_compiler/nodeset.py [hunk 3]
```diff
@@ -227,16 +227,16 @@ def filterRef(r, rt):
...
+                    self.nodes[r.source].references.items()
                 ))
             elif r.source == node.id:
                 if r.target not in self.nodes:
                     continue
-                self.nodes[r.target].references = set(filter(
-                    lambda rt: filterRef(r, rt),
-                    self.nodes[r.target].references
+                self.nodes[r.target].references = dict(filter(
+                    lambda rt: filterRef(r, rt[0]),
```
Comment (jpfr) [inline tools/nodeset_compiler/nodeset.py:238]: Why rt[0]?
Add a comment to document.

**Review by jpfr — COMMENTED** [review]
Looking good!
Good idea to just use dict.

Some minor comments. Otherwise good to go.

File: tools/nodeset_compiler/nodeset.py (hunk 2 above)
Comment (tobiajung) [reply tools/nodeset_compiler/nodeset.py]: Only the key is passed through the filter. The `dict.items` statement returns all key value pairs as views. In order to get the key, the first item must be selected.
I added a comment.

File: tools/nodeset_compiler/nodeset.py (hunk 3 above)
Comment (tobiajung) [reply tools/nodeset_compiler/nodeset.py]: only the key is passed through the filter. The dict.items statement returns all key, value pairs as views. In order to get the key, the first item must be selected.
I added a comment.

File: tools/nodeset_compiler/nodes.py (hunk 1 above)
Comment (tobiajung) [reply tools/nodeset_compiler/nodes.py]: Done

**Comment by tobiajung** [comment]
Seems like the CI failed.
I have no idea why. I only added some comments.
The log says that it has something to do with a missing mbedtls header:
```
[ 65%] Building C object CMakeFiles/open62541-object.dir/open62541.c.obj
D:\a\1\s\build\open62541.c:66008:10: fatal error: mbedtls/entropy_poll.h: No such file or directory
66008 | #include <mbedtls/entropy_poll.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
https://dev.azure.com/open62541/open62541/_build/results?buildId=5541&view=logs&j=8ae1b360-14bf-521d-a907-7a80f3a3a1f8&t=77a3e05d-c2b1-5f33-230d-a41770d4a825&l=299

**Comment by GoetzGoerisch** [comment]
@tobiajung I do have. It is now using mbetls 3.x

`Packages (1) mingw-w64-x86_64-mbedtls-3.1.0-1
`
https://dev.azure.com/open62541/open62541/_build/results?buildId=5541&view=logs&j=ad9e6d9e-6e5f-5424-e835-306608605e5b&t=2d5aafe1-b9f7-56bd-1478-b8e51f9d0a78&l=326

The mbedtls API was changed with v3.x, Needed changes are described in #4372

**Comment by jpfr** [comment]
Nice! Squash-merged.

**Pull request merged**